src/peppr/match.py

```diff
@@ -53,15 +55,85 @@ class UnmappableEntityError(Exception):
     pass
 
 
+class StructureMismatchError(Exception):
+    """
+    This exceptions is raised, if the reference and pose structure filtered to
+    ``matched`` atoms do not actually match.
+    This indicates some issue in the matching process.
+    """
+
+    pass
+
+
+def filter_matched(
+    reference: struc.AtomArray | struc.AtomArrayStack,
+    pose: struc.AtomArray | struc.AtomArrayStack,
+    prefilter: Callable[[struc.AtomArray], NDArray[np.bool_]] | None = None,
+) -> tuple[
+    struc.AtomArray | struc.AtomArrayStack, struc.AtomArray | struc.AtomArrayStack
+]:
+    """
+    Filter the matched atoms from the reference and pose, i.e.
+    where their ``matched`` annotation is ``True``.
+
+    Parameters
+    ----------
+    reference, pose : AtomArray
+        The structures to filter.
+        If they have a ``matched`` annotation, the atoms where ``matched`` is ``True``
+        are filtered.
+        Otherwise, all atoms are kept.
+    prefilter : Callable[AtomArray -> ndarray, dtype=bool], optional
+        For convenience, an additional filter function can be applied to both, the
+        reference and pose, before additionally filtering the matched atoms.
+
+    Returns
+    -------
+    matched_reference, matched_pose : AtomArray or AtomArrayStack
+        The filtered reference and pose.
+
+    Raises
+    ------
+    StructureMismatchError
+        If the matched structures do not have corresponding atoms, indicating an issue
+        in the prior matching process.
+    """
+    if "matched" in reference.get_annotation_categories():
+        ref_mask = reference.matched
+    else:
+        ref_mask = np.full(reference.array_length(), True)
+    if "matched" in pose.get_annotation_categories():
+        pose_mask = pose.matched
+    else:
+        pose_mask = np.full(pose.array_length(), True)
+    if prefilter is not None:
+        ref_mask = ref_mask & prefilter(reference)
+        pose_mask = pose_mask & prefilter(pose)
```

Comment:
admittedly personal preference
```suggestion
        ref_mask &= prefilter(reference)
        pose_mask& = prefilter(pose)
```